gcp-observability/src/main/java/io/grpc/gcp/observability/ObservabilityConfigImpl.java

```diff
@@ -74,95 +87,128 @@ void parse(String config) throws IOException {
+      parseMonitoringObject(config, CONFIG_MONITORING_VAR_NAME);
+      parseTracingObject(config, CONFIG_TRACING_VAR_NAME);
+      parseCustomTags(config, CONFIG_LABELS_VAR_NAME);
+    }
+  }
+
+  private void parseLoggingObject(Map<String, ?> config,  String jsonObjectName) {
+    Map<String, ?> rawCloudLogging = JsonUtil.getObject(config, jsonObjectName);
+    if (rawCloudLogging != null) {
+      enableCloudLogging = true;
+      List<?> rawClientRpcEvents = JsonUtil.getList(rawCloudLogging, "client_rpc_events");
+      clientLogFilters =
+          rawClientRpcEvents != null ? parseRpcEvents(rawClientRpcEvents) : clientLogFilters;
+      List<?> rawServerRpcEvents = JsonUtil.getList(rawCloudLogging, "server_rpc_events");
+      serverLogFilters =
+          rawServerRpcEvents != null ? parseRpcEvents(rawServerRpcEvents) : serverLogFilters;
+    }
+  }
+
+  private void parseMonitoringObject(Map<String, ?> config, String jsonObjectName) {
+    Map<String, ?> rawCloudMonitoring = JsonUtil.getObject(config, jsonObjectName);
+    if (rawCloudMonitoring != null) {
+      if (rawCloudMonitoring.isEmpty()) {
+        enableCloudMonitoring = true;
       }
-      rawList = JsonUtil.getList(config, "event_types");
-      if (rawList != null) {
-        List<String> jsonEventTypes = JsonUtil.checkStringList(rawList);
-        ImmutableList.Builder<EventType> eventTypesBuilder = new ImmutableList.Builder<>();
-        for (String jsonEventType : jsonEventTypes) {
-          eventTypesBuilder.add(convertEventType(jsonEventType));
+    }
+  }
+
+  private void parseTracingObject(Map<String, ?> config, String jsonObjectName) {
+    Map<String, ?> rawCloudTracing = JsonUtil.getObject(config, jsonObjectName);
+    if (rawCloudTracing != null) {
+      enableCloudTracing = true;
+      this.sampler = Samplers.probabilitySampler(0.0);
+      if (!rawCloudTracing.isEmpty()) {
+        Double samplingRate = JsonUtil.getNumberAsDouble(rawCloudTracing, "sampling_rate");
+        if (samplingRate != null) {
+          checkArgument(
+              samplingRate >= 0.0 && samplingRate <= 1.0,
+              "'sampling_rate' needs to be between [0.0, 1.0]");
+          // Using alwaysSample() instead of probabilitySampler() because according to
+          // {@link io.opencensus.trace.samplers.ProbabilitySampler#shouldSample}
+          // there is a (very) small chance of *not* sampling if probability = 1.00.
+          this.sampler =
+              1 - samplingRate < EPSILON
+                  ? Samplers.alwaysSample()
+                  : Samplers.probabilitySampler(samplingRate);
         }
-        this.eventTypes = eventTypesBuilder.build();
       }
-      Double samplingRate = JsonUtil.getNumberAsDouble(config, "global_trace_sampling_rate");
-      if (samplingRate == null) {
-        this.sampler = Samplers.probabilitySampler(0.0);
-      } else {
+    }
+  }
+
+  private void parseCustomTags(Map<String, ?> config, String jsonObjectName) {
```

Comment:
It will be good if all of your `parse*` methods follow this pattern:

- they take a `Map` argument. In this case the result of `JsonUtil.getObject(config, "labels")`
- they return a value which the caller assigns to something. In this case the caller assigns the returned value to `this.customTags`
- they are static since the method needs/uses no instance data with the above changes

Reply:
Done